pkg/currency/currency.go

```diff
@@ -0,0 +1,42 @@
+package currency
+
+import (
+	"errors"
+	"fmt"
+	"math"
+)
+
+// Picodollar is a type to represent currency with 12 decimal precision
+type PicoDollar int64
+
+const (
+	PicoDollarsPerDollar = 1e12
+)
+
+// FromDollars converts a dollar amount (as a float) to Picodollars
+// This should mostly be used for testing, and real usage should be done purely in PicoDollars
+func FromDollars(dollars float64) (PicoDollar, error) {
+	if math.IsNaN(dollars) || math.IsInf(dollars, 0) {
+		return 0, errors.New("invalid dollar amount: must be a finite number")
+	}
+
+	picodollars := dollars * PicoDollarsPerDollar
+	if picodollars > math.MaxInt64 || picodollars < math.MinInt64 {
```

Comment:
the first check wont work. because if it has overflown it wont be more than the max. It will be a negative number :)
Same applies to the min. It can never be less then the min.

You can either check that the dollars going into the function are small enough to not overflow, or check for overflow if the number became negative.

Dunno if dollars can be negative here.

Reply:
Fixed. That was dumb